plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+returnL (FrameAppArg argClosure   : ctx) heap cl                           = evaluateFun ctx heap cl argClosure
+returnL (FrameWrap ann tyn ty     : ctx) heap (Closure v env)              = returnL ctx heap $ Closure (Wrap ann tyn ty v) env
+returnL (FrameHeapUpdate l        : ctx) heap cl                           = returnL ctx heap' cl
+                                                                                   where heap' = updateHeap l (Evaluated cl) heap
+                                                                                      -- Leave this out to make the machine call-by-name (really slow)
+returnL (FrameUnwrap              : ctx) heap (Closure (Wrap _ _ _ t) env) = returnL ctx heap (Closure t env)
+returnL (FrameUnwrap              : _)      _ (Closure term _)             = throw $ LMachineException "Attemtping to unwrap non-wrapped term" term
+
+
+-- | Apply a function to an argument and proceed.
+-- If the function is a 'LamAbs', then extend the current environment with a new variable and proceed.
+-- If the function is not a 'LamAbs', then 'Apply' it to the argument and view this
+-- as an iterated application of a 'BuiltinName' to a list of 'Value's.
+-- If succesful, proceed with either this same term or with the result of the computation
+-- depending on whether 'BuiltinName' is saturated or not.
+
+evaluateFun :: EvaluationContext -> Heap -> Closure -> Closure-> LMachineResult
+evaluateFun ctx heap (Closure fun funEnv) argClosure =
+    case fun of
+      LamAbs _ var _ body ->
+          let (l, heap') = insertInHeap argClosure heap
+              env' = updateEnvironment (unUnique $ nameUnique var) l funEnv
+          in  computeL ctx heap' (Closure body env')
+
+      _ ->
+          -- Not a lambda: look for evaluation of a built-in function, possibly with some args already supplied.
+          -- We have to force the arguments, which means that we need to get the new heap back as well.
+          -- This bit is messy but should be much easier when we have n-ary application for builtins.
+          case computeL [] heap argClosure of
+            -- Force the argument, but only at the top level.
+            -- This is a bit of a hack.  We're not in the main compute/return process here.
+            -- We're preserving ctx, the context at entry, and we'll use that when we return.
+            -- We could also add a new type of stack frame for this.  Exactly what we do will
+            -- depend on the final interface to builtins.
+            Failure -> Failure
+            Success (Closure arg' env') heap' ->
+                let term = Apply () fun arg'
+                in case termAsPrimIterApp term of
+                     Nothing ->
+                         throw $ LMachineException "Trying to apply invalid term" term
+                         -- Was "Cannot reduce a not immediately reducible application."  This message isn't very helpful.
```

Comment:
Feel free to change the original error message in `Language.PlutusCore.Evaluation.MachineException`, but I like the previous one more, because it's more precise even though it's also more cryptic. In that case the term is not invalid, it's just of that the head of the application is not a `StagedBuiltinName` (i.e. either a `BuiltinName` or a `DynamicBuiltinName`). That's perhaps what the message should really say.